### PR TITLE
Fix maximum call stack size in indexMapper

### PR DESCRIPTION
**Pull request by budnix**
### Context

The PR fixes the RangeError (maximum call stack size exceeded) that happens when the Handsontable uses more hidden rows and/or columns than ~6000 records. I found that the error happens in two places. 
 * In the public `getFirstNotHiddenIndex` method of the _IndexMapper_ class;
 * In the private method `findVisibleCoordsInRowsRange` of the _VisualSelection_ class;

The methods used the complexity of O(n) where combined with recursion caused the methods to explode. For setup where all 6000 rows are hidden calling the API as `hot.getFirstNotHiddenIndex(0, 1)` causes about 6 thousand recursion calls until the method throws the RangeError.

~The PR rewrites the method by using a private index mapper cache used to translate from visual to renderable indexes. Thanks to that there is no necessary to make recursion calls. I observed the performance boost. The method executes several to several times faster than before. However, the method signature also is changed. I found that the method accepts arguments values that are not used at all and that unnecessarily increases the logic complexity that we need to maintain.~

The PR is not a breaking change. The original method will stay in the public API till the next major release. But for now, it's marked as `@deprecated` and it prints the warning message when used. The new bug-free and more simplified `getNearestNotHiddenIndex` method from now on taking a place of the original method.

The new method implements the logic of searching for the nearest non-hidden index way more performant but also cutting down the API to the minimum. 

#### Before
```ts
getFirstNotHiddenIndex(fromVisualIndex: number, incrementBy: number, searchAlsoOtherWayAround: boolean, indexForNextSearch: number);
```

#### After
```ts
getNearestNotHiddenIndex(fromVisualIndex: number, searchDirection: -1|1, searchAlsoOtherWayAround: boolean);
```

### How has this been tested?

I tested the changes locally. I added unit tests for a new method and smoke tests related to the bug linked to the PR.

### Types of changes

- [x] New feature or improvement (non-breaking change which adds functionality)

### Related issue(s):
1. #9064

### Affected project(s):
- [x] `handsontable`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [x] My change requires a change to the documentation.

### CODE REVIEW

**Comment by warpech** [comment]
It was agreed today that we will fix the problem and omit a sudden breaking change by doing the following:

* we will implement a new method `getNearestNotHiddenIndex` and use it everywhere in our code from now on
    * because of the new method, the new release will be a minor (12.1.0)
    * the new name actually fits the purpose of the method better
    * the PR should include input for the Release Notes to say that this method replaces the old method and why is it better
    * a minor release does NOT need a Migration Guide
* we will stop using the old method `getFirstNotHiddenIndex` in our code
    * the same PR should mark the old method as deprecated and present the alternative in the API reference
    * in the method body, we will write a `console.warn` (that executes only once), saying that you are using a deprecated method
    * the old method will be deleted in a future major release. In that major release, there would be a Migration Guide saying which method is the successor to `getFirstNotHiddenIndex`
    * the tests of the old method will remain until it is fully deleted

File: .changelogs/9064.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "title": "Fixed maximum call stack size error in indexMapper by introducing the new `getNearestNotHiddenIndex` method.",
```
Comment (kirszenbaum) [inline .changelogs/9064.json:2]:
```suggestion
  "title": "Fixed an issue where calling `getFirstNotHiddenIndex()` on a Handsontable instance with more than 6000 hidden rows or columns resulted in an uncaught `RangeError`, by adding a new method: `getNearestNotHiddenIndex()`. The `getFirstNotHiddenIndex()` method still works, but is marked as deprecated.
```

Comment (kirszenbaum) [reply .changelogs/9064.json:2]: @budnix – from what I see in the in the PR's description, we could also present it as "Added", like this:

```
  "title": "Added `getNearestNotHiddenIndex()`, a new method that finds a visual index of the nearest not-hidden row or column, and works even with large numbers of hidden rows or columns. The previous method, `getFirstNotHiddenIndex()` still works, but is marked as deprecated.
```

**Review by kirszenbaum — CHANGES_REQUESTED** [review]

File: .changelogs/9064.json (hunk 1 above)
Comment (budnix) [reply .changelogs/9064.json:2]: Thanks. I've created a new [PR](https://github.com/handsontable/handsontable/pull/9519) with your suggestions. I completely forgot to wait for your review 🤦

Comment (warpech) [reply .changelogs/9064.json:2]: > I completely forgot to wait for your review 🤦

You shouldn't wait, as we haven't changed the process. But it is completely fine and expected that @kirszenbaum reviews the docs asynchronously to the PR.

**Pull request merged**